sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -600,16 +627,56 @@ public void processElement(ProcessContext context) throws Exception {
         }
       }
 
+      public void processRecord(T record) throws RuntimeException {
+        try {
+          preparedStatement.clearParameters();
+          spec.getPreparedStatementSetter().setParameters(record, preparedStatement);
+          preparedStatement.addBatch();
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      }
+
       @FinishBundle
       public void finishBundle() throws Exception {
         executeBatch();
       }
 
-      private void executeBatch() throws SQLException {
+      private void executeBatch() throws SQLException, IOException, InterruptedException {
+        if (spec.isBackOffEnabled()) {
+          executeBatchWithBackOff();
+        } else {
+          if (batchCount > 0) {
+            preparedStatement.executeBatch();
+            connection.commit();
+            batchCount = 0;
+            records.clear();
+          }
+        }
+      }
+
+      private void executeBatchWithBackOff()
+              throws SQLException, IOException, InterruptedException {
         if (batchCount > 0) {
-          preparedStatement.executeBatch();
-          connection.commit();
+          Sleeper sleeper = Sleeper.DEFAULT;
+          BackOff backoff = BUNDLE_WRITE_BACKOFF.backoff();
+          while (true) {
+            // Batch upsert entities.
+            try {
+              int[] updates = preparedStatement.executeBatch();
+              connection.commit();
+              // Break if the commit threw no exception.
+              break;
+            } catch (SQLException exception) {
```

Comment:
Is the connection and the `preparedStatement` guaranteed to be in a good (reusable) state after any possible `SQLException`, either in `executeBatch` or in `commit`? Is `preparedStatement` guaranteed to be cleared?